platform/default/online_file_source.cpp

```diff
@@ -156,21 +159,16 @@ class OnlineFileSource::Impl {
         networkIsReachableAgain();
     }
 
-    void setMaximumConcurrentRequestsOverride(const uint32_t maximumConcurrentRequestsOverride_) {
-        maximumConcurrentRequestsOverride = maximumConcurrentRequestsOverride_;
+    uint32_t getMaximumConcurrentRequests() const {
+        return maximumConcurrentRequests;
     }
 
-private:
-
-    uint32_t getMaximumConcurrentRequests() const {
-        if (maximumConcurrentRequestsOverride > 0) {
-            return maximumConcurrentRequestsOverride;
-        }
-        else {
-            return HTTPFileSource::maximumConcurrentRequests();
-        }
+    void setMaximumConcurrentRequests(uint32_t maximumConcurrentRequests_) {
+        maximumConcurrentRequests = maximumConcurrentRequests_;
```

Comment:
Should this allow setting a `0` value?

Reply:
Hmmm. I'd see no reason why I would want to forbid setting it to `0` in the code. If for some reason the number is set to `0` (debug use cases? idk) I hope the person/code knows what is happening.

@asheemmamoowala  Would you prefer it being caught/invalidated? If not, I'd give it a merge 🙂

Reply:
When setting from `0` to a positive value, the `OnlineFileSource` does not automatically restart downloading requests. I think that is a bug.

Here's a test that would fail:

```
TEST(OnlineFileSource, TEST_REQUIRES_SERVER(ZeroMaxRequests)) {
    util::RunLoop loop;
    OnlineFileSource fs;
    std::size_t response_counter = 0;
    const std::size_t NUM_REQUESTS = 3;

    fs.setMaximumConcurrentRequests(0);

    // requesting a low priority resource
    Resource low_prio{ Resource::Unknown, "http://127.0.0.1:3000/load/1" };
    low_prio.setPriority(Resource::Priority::Low);
    std::unique_ptr<AsyncRequest> req_0 = fs.request(low_prio, [&](Response) {
        response_counter++;
        req_0.reset();
        EXPECT_EQ(response_counter, NUM_REQUESTS); // make sure this is responded last
        loop.stop();
    });

    // requesting two "regular" resources
    Resource regular1{ Resource::Unknown, "http://127.0.0.1:3000/load/2" };
    std::unique_ptr<AsyncRequest> req_1 = fs.request(regular1, [&](Response) {
        response_counter++;
        req_1.reset();
    });
    Resource regular2{ Resource::Unknown, "http://127.0.0.1:3000/load/3" };
    std::unique_ptr<AsyncRequest> req_2 = fs.request(regular2, [&](Response) {
        response_counter++;
        req_2.reset();
    });

    fs.setMaximumConcurrentRequests(1);

    loop.run();
}
```

Reply:
Thanks for this test @asheemmamoowala.

I've looked into this and the reason for this test to fail is not because requests are not picked up again but because the low priority request is processed before the other are. If the loop is not stopped, all three requests are processed, but the low priority one is processed first.

The low priority one is processed first, because the status is not offline and therefore the resources are not put in the queue at the same time and have no chance to be reordered - instead the first one that is passed is processed.

This test however passes:
```
TEST(OnlineFileSource, TEST_REQUIRES_SERVER(ZeroMaxRequests)) {
    util::RunLoop loop;
    OnlineFileSource fs;
    std::size_t response_counter = 0;
    const std::size_t NUM_REQUESTS = 3;

    fs.setMaximumConcurrentRequests(0);

    Resource res1{ Resource::Unknown, "http://127.0.0.1:3000/load/1" };
    std::unique_ptr<AsyncRequest> req1 = fs.request(res1, [&](Response) {
        response_counter++;
        req1.reset();
        // test if all requests arrive
        if (response_counter == NUM_REQUESTS) loop.stop();
    });
    Resource res2{ Resource::Unknown, "http://127.0.0.1:3000/load/2" };
    std::unique_ptr<AsyncRequest> req2 = fs.request(res2, [&](Response) {
        response_counter++;
        req2.reset();
        // test if all requests arrive
        if (response_counter == NUM_REQUESTS) loop.stop();
    });
    Resource res3{ Resource::Unknown, "http://127.0.0.1:3000/load/3" };
    std::unique_ptr<AsyncRequest> req3 = fs.request(res3, [&](Response) {
        response_counter++;
        req3.reset();
        // test if all requests arrive
        if (response_counter == NUM_REQUESTS) loop.stop();
    });


    fs.setMaximumConcurrentRequests(1);

    loop.run();
}
```

So I think this is unrelated to the question whether we'd want to allow an OnlineFileSource with zero maximum concurrent requests. Thinking about it, an `OnlineFileSource` with `0` doesn't make much sense. Should there be an exception thrown here - or should it just work as an OnlineFileSource with zero concurrent requests is supposed to work: not at all?

I'm leaning towards the later but that's mostly because I don't like the idea of throwing an exception or similar here.

Reply:
> Should there be an exception thrown here - or should it just work as an OnlineFileSource with zero concurrent requests is supposed to work: not at all?

You're right! Thanks for investigating my failed failing test. Lets keep this as is in that case.